rosbag2_compression/src/rosbag2_compression/zstd_compressor.cpp

```diff
@@ -0,0 +1,116 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "logging.hpp"
+#include "rosbag2_compression/zstd_compressor.hpp"
+
+namespace
+{
+// Increasing the compression level will:
+//   - Increase the time taken to compress
+//   - Decrease the size of the compressed data
+// Setting to zero uses Zstd's default value of 3.
+constexpr const int DEFAULT_ZSTD_COMPRESSION_LEVEL = 1;
+}
+
+namespace rosbag2_compression
+{
+
+std::unique_ptr<char[]> ZstdCompressor::get_input_buffer(const std::string & uri)
+{
+  try {
+    std::ifstream infile{uri};
+    infile.exceptions(std::ifstream::failbit);
+    // Get size and allocate
+    infile.seekg(0, std::ios::end);
+    const size_t decompressed_buffer_length = infile.tellg();
+    auto decompressed_buffer = std::make_unique<char[]>(decompressed_buffer_length);
+    // Go back and read in contents
+    infile.seekg(0 /* off */, std::ios::beg);
+    infile.read(decompressed_buffer.get(), decompressed_buffer_length);
+    return decompressed_buffer;
+  } catch (std::ios_base::failure & fail) {
+    std::cerr << "Caught IO stream exception while reading file: " << fail.what() << std::endl;
+    throw fail;
+  }
+}
+
+void ZstdCompressor::check_compression_result(const size_t & compression_result)
+{
+  if (ZSTD_isError(compression_result)) {
+    std::stringstream error;
+    error << "ZSTD compression error: " << ZSTD_getErrorName(compression_result);
+    throw std::runtime_error(error.str());
+  }
+  ROSBAG2_COMPRESSION_LOG_DEBUG("ZSTD compressed file.");
+}
+
+void ZstdCompressor::print_compression_statistics(
+  std::chrono::high_resolution_clock::time_point start,
+  std::chrono::high_resolution_clock::time_point end,
+  size_t decompressed_size, size_t compressed_size)
+{
+  const auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
+  const auto compression_ratio =
+    static_cast<double>(decompressed_size) / static_cast<double>(compressed_size);
+  ROSBAG2_COMPRESSION_LOG_DEBUG_STREAM(
+    "Compression statistics:\n" <<
+      "Time: " << duration.count() << " microseconds" <<
+      "Compression Ratio: " << compression_ratio
+  );
+}
+
+std::string ZstdCompressor::compress_uri(const std::string & uri)
+{
+  const auto start = std::chrono::high_resolution_clock::now();
+  const auto compressed_uri = uri + "." + get_compression_identifier();
+  try {
+    auto decompressed_buffer = get_input_buffer(uri);
+    auto decompressed_buffer_length = strlen(decompressed_buffer.get());
```

Comment:
`get_input_buffer` should return {unique_ptr, buffer_size}
Calling strlen requires that the input buffer is null terminated. Its also going to step through the entire buffer